src/dataflow/src/sink/tail.rs

```diff
@@ -71,19 +82,15 @@ pub fn tail<G>(
             results.sort_by_key(|(time, _)| *time);
             let mut results: Vec<Row> = results.into_iter().map(|(_, row)| row).collect();
 
-            if connector.emit_progress {
-                if let Some(upper) = batch_upper(batches.last()) {
-                    // The user has requested progress messages and there's at least one
-                    // batch. All of the batches might have zero rows, so we do not depend on
-                    // results at all. Another benefit of using upper (instead of the largest row
-                    // time) is that the batch's upper may be larger than the row time.
-                    packer.push(Datum::Decimal(Significand::new(i128::from(upper))));
-                    packer.push(Datum::True);
-                    // Fill in the diff column and all table columns with NULL.
-                    for _ in 0..(connector.object_columns + 1) {
-                        packer.push(Datum::Null);
-                    }
-                    results.push(packer.finish_and_reuse());
+            if let Some(batch) = batches.last() {
+                let progress_row = update_progress(
+                    &mut input_frontier,
+                    batch.desc.upper().borrow(),
+                    &mut packer,
+                    connector.object_columns + 1,
+                );
+                if connector.emit_progress && received_data && progress_row.is_some() {
```

Comment:
`received_data` will always be true here. Does it still need to be checked?

Reply:
Nope, I'll remove it. Sorry about that one, I meant to clean up the small things once the general shape of this was accepted.